Sometimes fails to activate LifecycleNodes automatically

We've found sometimes it fails to activate LifecycleNodes automatically. Manual activation is accepted successfully.
We are currently investigating this. Probably will try to replace the activation condition.

```sh
# Launch nodes
$ ros2 launch ros2_socketcan socket_can_bridge.launch.xml interface:=vcan0
[INFO] [launch]: All log files can be found below /home/kenji/.ros/log/2021-03-18-14-49-12-183689-KM-DAIV-69318
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [socket_can_receiver_node_exe-1]: process started with pid [69330]
[INFO] [socket_can_sender_node_exe-2]: process started with pid [69332]
[socket_can_receiver_node_exe-1] [INFO 1616046552.306117348] [socket_can_receiver]: interface: vcan0
[socket_can_receiver_node_exe-1] [INFO 1616046552.306181503] [socket_can_receiver]: interval(s): 0.010000
[socket_can_sender_node_exe-2] [INFO 1616046552.306349697] [socket_can_sender]: interface: vcan0
[socket_can_sender_node_exe-2] [INFO 1616046552.306387237] [socket_can_sender]: timeout(s): 0.010000
[socket_can_sender_node_exe-2] [WARN 1616046552.541547370] [rcl_lifecycle]: No transition matching 3 found for current state unconfigured
[ERROR] [launch_ros.actions.lifecycle_node]: Failed to make transition 'TRANSITION_ACTIVATE' for LifecycleNode '/socket_can_sender'
[socket_can_sender_node_exe-2] [ERROR 1616046552.541622919] []: Unable to start transition 3 from current state unconfigured: Transition is not registered., at /tmp/binarydeb/ros-foxy-rcl-lifecycle-1.1.10/src/rcl_lifecycle.c:350

# Check status
$ ros2 lifecycle get /socket_can_sender
inactive [2]

# Check receiver
$ ros2 lifecycle get /socket_can_receiver
active [3]

# Activate manually
$ ros2 lifecycle set /socket_can_sender activate
Transitioning successful

# Check status again
$ ros2 lifecycle get /socket_can_sender
active [3]
```

--- a/launch/socket_can_receiver.launch.py
+++ b/launch/socket_can_receiver.launch.py
@@ -16,11 +16,14 @@
 
 
 from launch import LaunchDescription
-from launch.actions import DeclareLaunchArgument, EmitEvent
+from launch.actions import (DeclareLaunchArgument, EmitEvent,
+                            RegisterEventHandler)
 from launch.conditions import IfCondition
+from launch.event_handlers import OnProcessStart
 from launch.events import matches_action
 from launch.substitutions import LaunchConfiguration, TextSubstitution
 from launch_ros.actions import LifecycleNode
+from launch_ros.event_handlers import OnStateTransition
 from launch_ros.events.lifecycle import ChangeState
 import lifecycle_msgs.msg
 
@@ -37,20 +40,36 @@ def generate_launch_description():
                                              }],
                                              output='screen')
 
-    socket_can_receiver_configure_trans_event = EmitEvent(
-        event=ChangeState(
-            lifecycle_node_matcher=matches_action(socket_can_receiver_node),
-            transition_id=lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE,
-        ),
-        condition=IfCondition(LaunchConfiguration('auto_configure')),
+    socket_can_receiver_configure_event_handler = RegisterEventHandler(
+        event_handler=OnProcessStart(
+            target_action=socket_can_receiver_node,
+            on_start=[
+                EmitEvent(
+                    event=ChangeState(
+                        lifecycle_node_matcher=matches_action(socket_can_receiver_node),
+                        transition_id=lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE,
+                    ),
+                    condition=IfCondition(LaunchConfiguration('auto_configure')),
+                ),
+            ],
+        )
     )
 
-    socket_can_receiver_activate_trans_event = EmitEvent(
-        event=ChangeState(
-            lifecycle_node_matcher=matches_action(socket_can_receiver_node),
-            transition_id=lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE,
+    socket_can_receiver_activate_event_handler = RegisterEventHandler(
+        event_handler=OnStateTransition(
+            target_lifecycle_node=socket_can_receiver_node,
+            start_state='configuring',
+            goal_state='inactive',
+            entities=[
+                EmitEvent(
+                    event=ChangeState(
+                        lifecycle_node_matcher=matches_action(socket_can_receiver_node),
+                        transition_id=lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE,
+                    ),
+                    condition=IfCondition(LaunchConfiguration('auto_activate')),
+                ),
+            ],
         ),
-        condition=IfCondition(LaunchConfiguration('auto_activate')),
     )
 
     return LaunchDescription([
@@ -59,6 +78,6 @@ def generate_launch_description():
         DeclareLaunchArgument('auto_configure', default_value='true'),
         DeclareLaunchArgument('auto_activate', default_value='true'),
         socket_can_receiver_node,
-        socket_can_receiver_configure_trans_event,
-        socket_can_receiver_activate_trans_event,
+        socket_can_receiver_configure_event_handler,
+        socket_can_receiver_activate_event_handler,
     ])

--- a/launch/socket_can_receiver.launch.py
+++ b/launch/socket_can_receiver.launch.py
@@ -25,20 +25,21 @@ from launch.substitutions import LaunchConfiguration, TextSubstitution
 from launch_ros.actions import LifecycleNode
 from launch_ros.event_handlers import OnStateTransition
 from launch_ros.events.lifecycle import ChangeState
-import lifecycle_msgs.msg
+from lifecycle_msgs.msg import Transition
 
 
 def generate_launch_description():
-    socket_can_receiver_node = LifecycleNode(package='ros2_socketcan',
-                                             executable='socket_can_receiver_node_exe',
-                                             name='socket_can_receiver',
-                                             namespace=TextSubstitution(text=''),
-                                             parameters=[{
-                                                 'interface': LaunchConfiguration('interface'),
-                                                 'interval_sec':
-                                                     LaunchConfiguration('interval_sec'),
-                                             }],
-                                             output='screen')
+    socket_can_receiver_node = LifecycleNode(
+        package='ros2_socketcan',
+        executable='socket_can_receiver_node_exe',
+        name='socket_can_receiver',
+        namespace=TextSubstitution(text=''),
+        parameters=[{
+            'interface': LaunchConfiguration('interface'),
+            'interval_sec':
+            LaunchConfiguration('interval_sec'),
+        }],
+        output='screen')
 
     socket_can_receiver_configure_event_handler = RegisterEventHandler(
         event_handler=OnProcessStart(
@@ -47,7 +48,7 @@ def generate_launch_description():
                 EmitEvent(
                     event=ChangeState(
                         lifecycle_node_matcher=matches_action(socket_can_receiver_node),
-                        transition_id=lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE,
+                        transition_id=Transition.TRANSITION_CONFIGURE,
                     ),
                     condition=IfCondition(LaunchConfiguration('auto_configure')),
                 ),
@@ -64,7 +65,7 @@ def generate_launch_description():
                 EmitEvent(
                     event=ChangeState(
                         lifecycle_node_matcher=matches_action(socket_can_receiver_node),
-                        transition_id=lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE,
+                        transition_id=Transition.TRANSITION_ACTIVATE,
                     ),
                     condition=IfCondition(LaunchConfiguration('auto_activate')),
                 ),

--- a/launch/socket_can_sender.launch.py
+++ b/launch/socket_can_sender.launch.py
@@ -25,20 +25,21 @@ from launch.substitutions import LaunchConfiguration, TextSubstitution
 from launch_ros.actions import LifecycleNode
 from launch_ros.event_handlers import OnStateTransition
 from launch_ros.events.lifecycle import ChangeState
-import lifecycle_msgs.msg
+from lifecycle_msgs.msg import Transition
 
 
 def generate_launch_description():
-    socket_can_sender_node = LifecycleNode(package='ros2_socketcan',
-                                           executable='socket_can_sender_node_exe',
-                                           name='socket_can_sender',
-                                           namespace=TextSubstitution(text=''),
-                                           parameters=[{
-                                               'interface': LaunchConfiguration('interface'),
-                                               'timeout_sec':
-                                                   LaunchConfiguration('timeout_sec'),
-                                           }],
-                                           output='screen')
+    socket_can_sender_node = LifecycleNode(
+        package='ros2_socketcan',
+        executable='socket_can_sender_node_exe',
+        name='socket_can_sender',
+        namespace=TextSubstitution(text=''),
+        parameters=[{
+            'interface': LaunchConfiguration('interface'),
+            'timeout_sec':
+            LaunchConfiguration('timeout_sec'),
+        }],
+        output='screen')
 
     socket_can_sender_configure_event_handler = RegisterEventHandler(
         event_handler=OnProcessStart(
@@ -47,7 +48,7 @@ def generate_launch_description():
                 EmitEvent(
                     event=ChangeState(
                         lifecycle_node_matcher=matches_action(socket_can_sender_node),
-                        transition_id=lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE,
+                        transition_id=Transition.TRANSITION_CONFIGURE,
                     ),
                     condition=IfCondition(LaunchConfiguration('auto_configure')),
                 ),
@@ -64,7 +65,7 @@ def generate_launch_description():
                 EmitEvent(
                     event=ChangeState(
                         lifecycle_node_matcher=matches_action(socket_can_sender_node),
-                        transition_id=lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE,
+                        transition_id=Transition.TRANSITION_ACTIVATE,
                     ),
                     condition=IfCondition(LaunchConfiguration('auto_activate')),
                 ),

--- a/launch/socket_can_sender.launch.py
+++ b/launch/socket_can_sender.launch.py
@@ -16,11 +16,14 @@
 
 
 from launch import LaunchDescription
-from launch.actions import DeclareLaunchArgument, EmitEvent
+from launch.actions import (DeclareLaunchArgument, EmitEvent,
+                            RegisterEventHandler)
 from launch.conditions import IfCondition
+from launch.event_handlers import OnProcessStart
 from launch.events import matches_action
 from launch.substitutions import LaunchConfiguration, TextSubstitution
 from launch_ros.actions import LifecycleNode
+from launch_ros.event_handlers import OnStateTransition
 from launch_ros.events.lifecycle import ChangeState
 import lifecycle_msgs.msg
 
@@ -37,20 +40,36 @@ def generate_launch_description():
                                            }],
                                            output='screen')
 
-    socket_can_sender_configure_trans_event = EmitEvent(
-        event=ChangeState(
-            lifecycle_node_matcher=matches_action(socket_can_sender_node),
-            transition_id=lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE,
-        ),
-        condition=IfCondition(LaunchConfiguration('auto_configure')),
+    socket_can_sender_configure_event_handler = RegisterEventHandler(
+        event_handler=OnProcessStart(
+            target_action=socket_can_sender_node,
+            on_start=[
+                EmitEvent(
+                    event=ChangeState(
+                        lifecycle_node_matcher=matches_action(socket_can_sender_node),
+                        transition_id=lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE,
+                    ),
+                    condition=IfCondition(LaunchConfiguration('auto_configure')),
+                ),
+            ],
+        )
     )
 
-    socket_can_sender_activate_trans_event = EmitEvent(
-        event=ChangeState(
-            lifecycle_node_matcher=matches_action(socket_can_sender_node),
-            transition_id=lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE,
+    socket_can_sender_activate_event_handler = RegisterEventHandler(
+        event_handler=OnStateTransition(
+            target_lifecycle_node=socket_can_sender_node,
+            start_state='configuring',
+            goal_state='inactive',
+            entities=[
+                EmitEvent(
+                    event=ChangeState(
+                        lifecycle_node_matcher=matches_action(socket_can_sender_node),
+                        transition_id=lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE,
+                    ),
+                    condition=IfCondition(LaunchConfiguration('auto_activate')),
+                ),
+            ],
         ),
-        condition=IfCondition(LaunchConfiguration('auto_activate')),
     )
 
     return LaunchDescription([
@@ -59,6 +78,6 @@ def generate_launch_description():
         DeclareLaunchArgument('auto_configure', default_value='true'),
         DeclareLaunchArgument('auto_activate', default_value='true'),
         socket_can_sender_node,
-        socket_can_sender_configure_trans_event,
-        socket_can_sender_activate_trans_event,
+        socket_can_sender_configure_event_handler,
+        socket_can_sender_activate_event_handler,
     ])

